`mpmath/tests/test_convert.py::test_compatibility` fails on ppc64, sparc (possibly more)

While testing on PPC64 and SPARC, we've noticed the following test failure that seems to be related to longdouble precision:

```pytb
$ python -m pytest mpmath/tests/test_convert.py::test_compatibility  -vv
========================================================= test session starts =========================================================
platform linux -- Python 3.12.3, pytest-8.3.2, pluggy-1.5.0 -- /home/mgorny/mpmath/.venv/bin/python
mpmath backend: python
mpmath mp class: <mpmath.ctx_mp.MPContext object at 0x3fff9c72fe60>
mpmath version: 1.4.0a2.dev67+gcf98bcf
Python version: 3.12.3 (main, May 31 2024, 13:52:31) [GCC 12.3.1 20230526]
cachedir: .pytest_cache
rootdir: /home/mgorny/mpmath
configfile: pyproject.toml
plugins: xdist-3.6.1
collected 1 item                                                                                                                      

mpmath/tests/test_convert.py::test_compatibility FAILED                                                                         [100%]

============================================================== FAILURES ===============================================================
_________________________________________________________ test_compatibility __________________________________________________________

    def test_compatibility():
        from packaging.version import Version, parse
        np = pytest.importorskip("numpy")
        if parse(np.__version__) < Version('2.0.0b1'):
            npcore = np.core
        else:
            npcore = np._core
        # numpy types
        for nptype in npcore.numerictypes.typeDict.values():
            if issubclass(nptype, np.complexfloating):
                x = nptype(complex(0.5, -0.5))
            elif issubclass(nptype, np.floating):
                x = nptype(0.5)
            elif issubclass(nptype, np.integer):
                x = nptype(2)
            # Handle the weird types
            try: diff = np.abs(type(np.sqrt(x))(sqrt(x)) - np.sqrt(x))
            except: continue
            assert diff < np.float64(2.0**-53)
        assert mpf(np.float64('inf')) == inf
        assert isnan(mp.npconvert(np.float64('nan')))
        if hasattr(np, "float128"):
            mp.prec = 113
>           assert (mp.npconvert(np.float128('0.841470984807896506652502321630298954')) ==
                    mpf('0.841470984807896506664590813295845351'))
E           AssertionError: assert mpf('0.841470984807896506652502321630296065') == mpf('0.841470984807896506664590813295845351')
E            +  where mpf('0.841470984807896506652502321630296065') = npconvert(np.longdouble('0.8414709848078965066525023216303'))
E            +    where npconvert = <mpmath.ctx_mp.MPContext object at 0x3fff9c72fe60>.npconvert
E            +    and   np.longdouble('0.8414709848078965066525023216303') = <class 'numpy.longdouble'>('0.841470984807896506652502321630298954')
E            +      where <class 'numpy.longdouble'> = <module 'numpy' from '/home/mgorny/mpmath/.venv/lib/python3.12/site-packages/numpy/__init__.py'>.float128
E            +  and   mpf('0.841470984807896506664590813295845351') = mpf('0.841470984807896506664590813295845351')

mpmath/tests/test_convert.py:267: AssertionError
======================================================== slowest 20 durations =========================================================
0.14s call     mpmath/tests/test_convert.py::test_compatibility
0.00s setup    mpmath/tests/test_convert.py::test_compatibility
0.00s teardown mpmath/tests/test_convert.py::test_compatibility
======================================================= short test summary info =======================================================
FAILED mpmath/tests/test_convert.py::test_compatibility - AssertionError: assert mpf('0.841470984807896506652502321630296065') == mpf('0.841470984807896506664590813295845351')
 +  where mpf('0.841470984807896506652502321630296065') = npconvert(np.longdouble('0.8414709848078965066525023216303'))
 +    where npconvert = <mpmath.ctx_mp.MPContext object at 0x3fff9c72fe60>.npconvert
 +    and   np.longdouble('0.8414709848078965066525023216303') = <class 'numpy.longdouble'>('0.841470984807896506652502321630298954')
 +      where <class 'numpy.longdouble'> = <module 'numpy' from '/home/mgorny/mpmath/.venv/lib/python3.12/site-packages/numpy/__init__.py'>.float128
 +  and   mpf('0.841470984807896506664590813295845351') = mpf('0.841470984807896506664590813295845351')
========================================================== 1 failed in 0.34s ==========================================================
```

To reproduce (on PPC64 or SPARC):

```
pip install -e . pytest numpy
python -m pytest mpmath/tests/test_convert.py::test_compatibility
```

Not sure if it's a matter of better or worse precision, though.

Hmm, I think not.

ppc64:

```
FAILED mpmath/tests/test_convert.py::test_compatibility - AssertionError: assert mpf('0.841470984807896506652502321630296065') == mpf('0.841470984807896506664590813295845351')
```

sparc:
```
FAILED mpmath/tests/test_convert.py::test_compatibility - AssertionError: assert mpf('0.841470984807896506652502321630298954') == mpf('0.841470984807896506664590813295845351')
```

AArch64:

```
FAILED mpmath/tests/test_convert.py::test_compatibility - AssertionError: assert mpf('0.841470984807896506652502321630298954') == mpf('0.841470984807896506664590813295845351')
```

Further results:

PA-RISC:
```
FAILED mpmath/tests/test_convert.py::test_compatibility - AttributeError: module 'numpy' has no attribute 'float128'. Did you mean: 'float16'?
```

MIPS64:
```
FAILED mpmath/tests/test_convert.py::test_compatibility - AssertionError: assert mpf('0.841470984807896506652502321630298954') == mpf('0.841470984807896506664590813295845351')
```

Alpha:
```
FAILED mpmath/calculus/optimization.py::mpmath.calculus.optimization.steffensen
FAILED mpmath/tests/test_convert.py::test_compatibility - AssertionError: assert mpf('0.841470984807896506652502321630298954') == mpf('0.841470984807896506664590813295845351')
```

The other failure from alpha is:
```pytb
_____________ [doctest] mpmath.calculus.optimization.steffensen ______________
1034 
1035     x* is a fixpoint of the iteration x_{k+1} = phi(x_k) if x* = phi(x*). For
1036     phi(x) = x**2 there are two fixpoints: 0 and 1.
1037 
1038     Let's try Steffensen's method:
1039 
1040     >>> f = lambda x: x**2
1041     >>> from mpmath.calculus.optimization import steffensen
1042     >>> F = steffensen(f)
1043     >>> for x in [0.5, 0.9, 2.0]:
Differences (unified diff with -expected +actual):
    @@ -5,5 +5,5 @@
              2.32831e-10          -2.84465e-27
              5.42101e-20           2.30189e-80
    -         2.93874e-39          -1.2197e-239
    +         2.93874e-39                     0
              8.63617e-78                     0
             7.45834e-155                     0
```

> FAILED mpmath/tests/test_convert.py::test_compatibility - AttributeError: module 'numpy' has no attribute 'float128'. Did you mean: 'float16'?

Here does make sense traceback.

BTW, what if you change mp.prec on L266 to 64?

> FAILED mpmath/calculus/optimization.py::mpmath.calculus.optimization.steffensen

This is another issue, probably related with fp backend.  Could you open a new bug, please?

> > FAILED mpmath/tests/test_convert.py::test_compatibility - AttributeError: module 'numpy' has no attribute 'float128'. Did you mean: 'float16'?
> 
> Here does make sense traceback.

Ah, sorry. Since I've asked someone to test this for us, they've tested alpha1. I see a `hasattr()` has been added past it, so please disregard this.

> BTW, what if you change mp.prec on L266 to 64?

If this is about ppc64, I'm getting then:

```
FAILED mpmath/tests/test_convert.py::test_compatibility - AssertionError: assert mpf('0.841470984807896506653') == mpf('0.841470984807896506665')
```

So it's close but still slightly off.

> 
> > FAILED mpmath/calculus/optimization.py::mpmath.calculus.optimization.steffensen
> 
> This is another issue, probably related with fp backend. Could you open a new bug, please?

I will, thanks.

No, 64 doesn't work there. AArch64 and SPARC both give:

```
FAILED mpmath/tests/test_convert.py::test_compatibility - AssertionError: assert mpf('0.841470984807896506653') == mpf('0.841470984807896506665')
```

No, I did not change RHS. If I change it like suggested, all ppc64le, aarch64 and sparc fail this way:

```
E           AssertionError: assert mpf('0.841470984807896506653') == mpf('0.841470984807896506665')
```

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -39,7 +39,7 @@ jobs:
     - name: Install dependencies
       run: |
         pip install --upgrade setuptools pip
-        pip install --upgrade .[develop,gmpy,docs,ci]
+        pip install --upgrade .[develop,gmpy2,docs,ci]
     - name: Remove gmpy (for coverage tests)
       if: matrix.nogmpy
       run: pip uninstall -y gmpy2

--- a/docs/setup.rst
+++ b/docs/setup.rst
@@ -19,10 +19,10 @@ or some specific version with::
 
     pip install mpmath==1.3.0
 
-You can install also extra dependencies, e.g. `gmpy
+You can install also extra dependencies, e.g. `gmpy2
 <https://github.com/aleaxit/gmpy>`_ support::
 
-    pip install mpmath[gmpy]
+    pip install mpmath[gmpy2]
 
 .. tip::
 
@@ -81,18 +81,18 @@ Python interpreter and do the following::
        python -m mpmath
 
 
-Using gmpy (optional)
----------------------
+Using gmpy2 (optional)
+----------------------
 
-If `gmpy <https://github.com/aleaxit/gmpy>`_ version 2.2.0 or later is
+If `gmpy2 <https://github.com/aleaxit/gmpy>`_ version 2.2.0 or later is
 installed on your system, mpmath will automatically detect it and transparently
-use gmpy integers instead of Python integers.  This makes mpmath much faster,
+use gmpy2 integers instead of Python integers.  This makes mpmath much faster,
 especially at high precision (approximately above 100 digits).
 
-To verify that mpmath uses gmpy, check the internal variable ``BACKEND`` is
+To verify that mpmath uses gmpy2, check the internal variable ``BACKEND`` is
 equal to 'gmpy'.
 
-Using the gmpy backend can be disabled by setting the ``MPMATH_NOGMPY``
+Using the gmpy2 backend can be disabled by setting the ``MPMATH_NOGMPY``
 environment variable.  Note that the mode cannot be switched during runtime;
 mpmath must be re-imported for this change to take effect.
 

--- a/mpmath/calculus/optimization.py
+++ b/mpmath/calculus/optimization.py
@@ -1038,10 +1038,11 @@ def steffensen(f):
     Let's try Steffensen's method:
 
     >>> f = lambda x: x**2
+    >>> from mpmath import mp
     >>> from mpmath.calculus.optimization import steffensen
     >>> F = steffensen(f)
     >>> for x in [0.5, 0.9, 2.0]:
-    ...     fx = Fx = x
+    ...     fx = Fx = mp.mpf(x)
     ...     for i in range(9):
     ...         try:
     ...             fx = f(fx)
@@ -1051,16 +1052,16 @@ def steffensen(f):
     ...             Fx = F(Fx)
     ...         except ZeroDivisionError:
     ...             pass
-    ...         print('%20g  %20g' % (fx, Fx))
+    ...         print(f'{fx:20g}  {Fx:20g}')
                     0.25                  -0.5
                   0.0625                   0.1
               0.00390625            -0.0011236
              1.52588e-05           1.41691e-09
              2.32831e-10          -2.84465e-27
              5.42101e-20           2.30189e-80
-             2.93874e-39          -1.2197e-239
-             8.63617e-78                     0
-            7.45834e-155                     0
+             2.93874e-39         -1.21971e-239
+             8.63617e-78          1.81455e-717
+            7.45834e-155        -5.97459e-2151
                     0.81               1.02676
                   0.6561               1.00134
                 0.430467                     1

--- a/mpmath/ctx_mp_python.py
+++ b/mpmath/ctx_mp_python.py
@@ -704,7 +704,7 @@ class PythonMPContext:
         import numpy as np
         if isinstance(x, np.ndarray) and x.ndim == 0: x = x.item()
         if isinstance(x, (np.integer, int)): return ctx.make_mpf(from_int(int(x)))
-        if isinstance(x, (np.floating, float)): return ctx.make_mpf(from_npfloat(x))
+        if isinstance(x, (np.floating, float)): return ctx.mpf(from_npfloat(x))
         if isinstance(x, (np.complexfloating, complex)):
             return ctx.make_mpc((from_npfloat(x.real), from_npfloat(x.imag)))
         raise TypeError("cannot create mpf from " + repr(x))

--- a/mpmath/functions/zeta.py
+++ b/mpmath/functions/zeta.py
@@ -288,6 +288,10 @@ def bernpoly(ctx, n, z):
     n = int(n)
     if n < 0:
         raise ValueError("Bernoulli polynomials only defined for n >= 0")
+    if ctx.isinf(z):
+        return z ** n
+    if ctx.isnan(z):
+        return z
     if n <= 3:
         if n == 0: return z ** 0
         if n == 1: return z - 0.5
@@ -297,10 +301,6 @@ def bernpoly(ctx, n, z):
         return ctx.bernoulli(n)
     if z == 0.5:
         return (ctx.ldexp(1,1-n)-1)*ctx.bernoulli(n)
-    if ctx.isinf(z):
-        return z ** n
-    if ctx.isnan(z):
-        return z
     if abs(z) > 2:
         def terms():
             t = ctx.one

--- a/mpmath/libmp/libmpc.py
+++ b/mpmath/libmp/libmpc.py
@@ -260,7 +260,7 @@ def mpc_pow_int(z, n, prec, rnd=round_fast):
     de = aexp - bexp
     abs_de = abs(de)
     exact_size = n*(abs_de + max(abc, bbc))
-    if exact_size < 10000:
+    if exact_size < 10000 and min(abc, bbc) >= 0:
         if de > 0:
             aman <<= de
             aexp = bexp

--- a/mpmath/libmp/libmpf.py
+++ b/mpmath/libmp/libmpf.py
@@ -3,14 +3,10 @@ Low-level functions for arbitrary-precision floating-point arithmetic.
 """
 
 import math
+import random
 import re
 import sys
 import warnings
-
-
-# Importing random is slow
-#from random import getrandbits
-getrandbits = None
 
 from .backend import BACKEND, MPZ, MPZ_FIVE, MPZ_ONE, MPZ_ZERO, gmpy
 from .libintmath import (bctable, bin_to_radix, isqrt, numeral, sqrtrem,
@@ -444,11 +440,7 @@ def to_fixed(s, prec):
 def mpf_rand(prec):
     """Return a raw mpf chosen randomly from [0, 1), with prec bits
     in the mantissa."""
-    global getrandbits
-    if not getrandbits:
-        import random
-        getrandbits = random.getrandbits
-    return from_man_exp(getrandbits(prec), -prec, prec, round_floor)
+    return from_man_exp(random.getrandbits(prec), -prec, prec, round_floor)
 
 def mpf_eq(s, t):
     """Test equality of two raw mpfs. This is simply tuple comparison

--- a/mpmath/libmp/libmpi.py
+++ b/mpmath/libmp/libmpi.py
@@ -2,11 +2,12 @@
 Computational functions for interval arithmetic.
 
 """
+from .backend import MPZ_ONE
 from .gammazeta import mpc_loggamma, mpf_gamma, mpf_loggamma, mpf_rgamma
 from .libelefun import (mod_pi2, mpf_atan, mpf_atan2, mpf_cos_sin, mpf_exp,
                         mpf_log, mpf_pi, mpf_sqrt)
-from .libmpf import (MPZ_ONE, dps_to_prec, fhalf, finf, fnan, fninf, fnone,
-                     fone, from_float, from_int, from_man_exp, from_str, fzero,
+from .libmpf import (dps_to_prec, fhalf, finf, fnan, fninf, fnone, fone,
+                     from_float, from_int, from_man_exp, from_str, fzero,
                      mpf_abs, mpf_add, mpf_div, mpf_ge, mpf_gt, mpf_le, mpf_lt,
                      mpf_min_max, mpf_mul, mpf_neg, mpf_pos, mpf_pow_int,
                      mpf_shift, mpf_sign, mpf_sub, prec_to_dps, round_ceiling,

--- a/mpmath/tests/test_convert.py
+++ b/mpmath/tests/test_convert.py
@@ -263,9 +263,9 @@ def test_compatibility():
     assert mpf(np.float64('inf')) == inf
     assert isnan(mp.npconvert(np.float64('nan')))
     if hasattr(np, "float128"):
-        mp.prec = 113
+        mp.prec = 64
         assert (mp.npconvert(np.float128('0.841470984807896506652502321630298954')) ==
-                mpf('0.841470984807896506664590813295845351'))
+                mpf('0.841470984807896506653'))
     mp.prec = 53
     # issues 382 and 539
     assert mp.sqrt(np.int64(1)) == mpf('1.0')

--- a/mpmath/tests/test_gammazeta.py
+++ b/mpmath/tests/test_gammazeta.py
@@ -694,3 +694,7 @@ def test_issue_471():
     assert bernpoly(4, inf) == inf
     assert bernpoly(4, mpc(inf, 0)) == mpc(inf, 0)
     assert isnan(bernpoly(4, nan))
+
+def test_issue_472():
+    assert bernpoly(4, mpc(inf, 1e-50)) == mpc(inf, 0)
+    assert mpc(inf, 2)**4 == mpc(inf, 0)

--- a/mpmath/tests/test_gammazeta.py
+++ b/mpmath/tests/test_gammazeta.py
@@ -689,3 +689,8 @@ def test_issue_723():
     mp.dps = 16
     assert zeta(-0.01 + 1000j).ae(-8.971459529241107 + 8.732179332810066j)
     mp.dps = 15
+
+def test_issue_471():
+    assert bernpoly(4, inf) == inf
+    assert bernpoly(4, mpc(inf, 0)) == mpc(inf, 0)
+    assert isnan(bernpoly(4, nan))

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -30,7 +30,7 @@ Homepage = 'https://mpmath.org/'
 'Bug Tracker' = 'https://github.com/mpmath/mpmath/issues'
 Documentation = 'http://mpmath.org/doc/current/'
 [project.optional-dependencies]
-tests = ['pytest>=6', 'numpy; python_version<"3.13"', 'packaging',
+tests = ['pytest>=6', 'numpy<2.1; python_version<"3.13"', 'packaging',
          'matplotlib; python_version<"3.13" and platform_python_implementation!="PyPy"',
          'pexpect', 'ipython', 'hypothesis']
 develop = ['mpmath[tests]', 'flake518>=1.5; python_version>="3.9"',

